### PR TITLE
FK & UNQ Constraint Violations

### PR DESCRIPTION
**Pull request by Hydrocharged**
Implements foreign key and unique key constraint violations.
Tests may be found at https://github.com/dolthub/dolt/pull/1870

### CODE REVIEW

File: go/libraries/doltcore/merge/constraint_violations.go
Comment (bheni) [inline go/libraries/doltcore/merge/constraint_violations.go:153]: Add a method to schema.Index so you can just do index.ToTableTuple(nbf, indexKey) or something?

File: go/libraries/doltcore/merge/constraint_violations.go
Comment (bheni) [inline go/libraries/doltcore/merge/constraint_violations.go:32]: There are a lot of objects used in this calculation.  For example:

parentFkConstraintViolations(
				ctx,
				foreignKey,
				postChildTable,
				preParentSch, postParentSch, postChildSch,
				postParentIndex, postChildIndex,
				preParentData, postParentData, postChildData, postParentIndexData, postChildIndexData)

Could possibly clean this up and get rid of some code duplication by creating a class that groups some of these elements together.

File: go/libraries/doltcore/merge/constraint_violations.go
Comment (bheni) [inline go/libraries/doltcore/merge/constraint_violations.go:191]: This function is a monster.  Could use some refactoring to improve readability.

File: go/libraries/doltcore/merge/constraint_violations.go
Comment (bheni) [inline go/libraries/doltcore/merge/constraint_violations.go:266]: Can put this all in an anonymous function to fix your defer call

File: go/libraries/doltcore/merge/constraint_violations.go
Comment (bheni) [inline go/libraries/doltcore/merge/constraint_violations.go:285]: extract this into a method to fix the defer close issue?  Also this function is a beast so could also help with clarity to pull this out.

**Review by bheni — COMMENTED** [review]
Looks pretty good.  Gonna need lots of testing to validate that the logic is correct.

**Review by andrew-wm-arthur — COMMENTED** [review]
This is still merging the primary row storage using `SessionedTableEditor`, which applies reference options when the rows are merged. 
Reference options must be considered after the primary row storage is merged in order to be applied correctly

**Comment by Hydrocharged** [comment]
> 
> 
> This is still merging the primary row storage using `SessionedTableEditor`, which applies reference options when the rows are merged.
> Reference options must be considered after the primary row storage is merged in order to be applied correctly

Foreign keys are disabled while merging.
https://github.com/dolthub/dolt/blob/master/go/libraries/doltcore/merge/merge.go#L793

**Comment by andrew-wm-arthur** [comment]
> > This is still merging the primary row storage using `SessionedTableEditor`, which applies reference options when the rows are merged.
> > Reference options must be considered after the primary row storage is merged in order to be applied correctly
> 
> Foreign keys are disabled while merging the primary row storage.
> https://github.com/dolthub/dolt/blob/151ca17cb21f782a3bcb392423dcdedc4e10d573/go/libraries/doltcore/merge/merge.go#L792-L794

My mistake, I didn't realize we were already ignoring reference options on merge

**Review by bheni — APPROVED** [review]
LGTM after you merge your bats tests.  Minor comments.

File: go/cmd/dolt/commands/merge.go
Comment (Hydrocharged) [reply go/cmd/dolt/commands/merge.go:137]: Working on that now, will be in a separate PR.

File: go/libraries/doltcore/env/actions/staged.go
Comment (Hydrocharged) [reply go/libraries/doltcore/env/actions/staged.go:113]: Good catch!

**Pull request merged**
